parmed/openmm/parameters.py

```diff
@@ -358,71 +413,33 @@ def get_types(residue, atomname):
-                                        warnings.warn("{} and {} match improper {}. Using {}".format(key, key_placeholder,
-                                                      altkeys1, key_placeholder))
-                                        break
-
-                                    if flag[0] == key[0] and flag[1] == key[1]:
-                                        improper_harmonic[altkeys1] = self.improper_types[key]
-                                        warnings.warn("{} and {} match improper {}. Using {}".format(key, key_placeholder,
-                                                        altkeys1, key), ParameterWarning)
-
-                                MATCH = True
-                                key_placeholder = key
-                                improper_harmonic[altkeys1] = self.improper_types[key]
-                            if key not in self.improper_types:
-                                for anchor in itertools.combinations([t1, t2, t3, t4], 2):
-                                    key = tuple(sorted([anchor[0], anchor[1], 'X', 'X']))
-                                    if key in self.improper_periodic_types:
-                                        if MATCH and key != key_placeholder:
-                                            flag = (altkeys1[0], altkeys1[-1])
-                                            if flag[0] == key_placeholder[0] and flag[1] == key_placeholder[1]:
-                                                # Match already found.
-                                                warnings.warn("{} and {} match improper {}. Using {}".format(key,
-                                                              key_placeholder, altkeys1, key_placeholder), ParameterWarning)
-                                                break
-
-                                            if flag[0] == key[0] and flag[1] == key[1]:
-                                                warnings.warn("More than one improper matches for {}. Using {}".format(
-                                                        altkeys1, key), ParameterWarning)
-                                                improper_periodic[altkeys1] = self.improper_periodic_types[key]
-
-                                        MATCH = True
-                                        key_placeholder = key
-                                        improper_periodic[altkeys1] = self.improper_periodic_types[key]
-                    if not MATCH:
-                        warnings.warn("No improper parameter found for {}".format(altkeys1), ParameterWarning)
+                        raise Exception('Something went wrong with improper type for {} returning an unexpected object {}'.format(key, improper))
+
+                # Warn if no improper was found
+                if not improper_found:
+                    warnings.warn('No improper found for improper {} in residue {} (types were {})'.format(impr, name, types))
```

Comment:
Should this have an explicit warning type other than `UserWarning` so it can be turned into a fatal error if a missing improper type is detected? What’s the use case for this not just raising an exception?

Reply:
Fixed.